hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/CryptoUpdateHandler.java

```diff
@@ -268,13 +269,19 @@ private void validateSemantics(
         // validate auto associations
         if (op.hasMaxAutomaticTokenAssociations()) {
             final long newMax = builderAccount.maxAutoAssociations();
+            // first validate if the associations are limited and the new max is within
             validateFalse(
-                    newMax < updateAccount.usedAutoAssociations(), EXISTING_AUTOMATIC_ASSOCIATIONS_EXCEED_GIVEN_LIMIT);
-            validateFalse(
-                    newMax > ledgerConfig.maxAutoAssociations(),
+                    entitiesConfig.limitTokenAssociations() && newMax > tokensConfig.maxPerAccount(),
                     REQUESTED_NUM_AUTOMATIC_ASSOCIATIONS_EXCEEDS_ASSOCIATION_LIMIT);
             validateFalse(
-                    entitiesConfig.limitTokenAssociations() && newMax > tokensConfig.maxPerAccount(),
+                    newMax < -1 && entitiesConfig.unlimitedAutoAssociationsEnabled()
+                            || newMax < 0 && !entitiesConfig.unlimitedAutoAssociationsEnabled(),
+                    INVALID_MAX_AUTO_ASSOCIATIONS);
+            validateFalse(
+                    newMax < updateAccount.usedAutoAssociations() && newMax != -1,
+                    EXISTING_AUTOMATIC_ASSOCIATIONS_EXCEED_GIVEN_LIMIT);
+            validateFalse(
+                    newMax > ledgerConfig.maxAutoAssociations(),
```

Comment:
Should this also check newMax != -1 ?

Reply:
The above validation is covering the case with -1.